### ISSUE TITLE
No way to freeze fused BN stats

### ISSUE THREAD

**Issue by DeepLearningJohnDoe** [tensorflow/tensorflow#10857]
When fine-tuning networks trained with BN sometimes we want to freeze and use the accumulated moving averages while allowing the gradients to be backpropagated through the BN layer, but currently there is no way of doing so with fused BN, since when is_training = False the layer gives erroneous gradients. Of course, we could use the batch statistics from the new task to accumulate the stats, but it isn't possible in the case of batch_size = 1.

I understand that due to the nature of the CuDNN kernel it might be hard to implement such feature, but a fused Batch Renorm layer could be a decent compromise, as it uses the moving averages when training as well as during inference.

**Comment by tatatodd** [comment]
@zhangyaobit can you comment on this?  Thanks!

**Comment by zhangyaobit** [comment]
Setting is_training to False and at the same time doing gradient computation is expected to get erroneous gradients, because the second and third outputs of FusedBatchNorm are only applicable/valid when is_training is True. 

Could you try train with fused batch norm, and fine-tuning with non-fused batch norm?

**Comment by DeepLearningJohnDoe** [comment]
In object detection and semantic segmentation sometimes the batch_size is set to 1 for the original network, and unfortunately our use case fall under this category. Non-fused batch norm for NCHW is around 6x slower https://github.com/tensorflow/tensorflow/issues/7551#issuecomment-280421351.

**Comment by zhangyaobit** [comment]
Not sure if I understand correctly: are you saying fine tune is done with a batch size of 1 and non-fused batch norm doesn't work with a batch size of 1? And even if it works, it is 6x slower?

How long do you expect fine tune currently takes (on the scale of hours or days? say on GPU)?

**Comment by DeepLearningJohnDoe** [comment]
Yes, fine tune is done with a batch size of 1 (on each GPU). I am training on a 4x 1080 machine, and fine tuning would take around ~20 hours. Non-fused batch norm does work (gives correct gradients), but it is way too slow and will push the fine tuning time beyond 100 hours. It would be faster to simply retrain the original network in NHWC and fine tune there with non-fused batch norm, which is ironic, considering NCHW is the CuDNN canonic format.

**Comment by zhangyaobit** [comment]
It may make sense to simply forward the population mean and variance to the second and third outputs of FusedBatchNorm when fused=Tue. There is no performance penalty for this, since it is a forwarding instead of a copy.

Before this feature is implemented, you can modify the code [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_impl.py#L815)  to set batch_mean and batch_variance to mean and variance when is_training is False. Note that doing this will introduce actual copies (a slight performance penalty), so as ideal as making this change on the C++ side; but this should get you most of the GPU performance.

**Comment by zhangyaobit** [comment]
I marked this as contribution welcome for now. Anyone on the Tensorflow team or external developers are welcome to contribute if interested. Please let me know if you have any questions.

**Comment by zhangyaobit** [comment]
I'm closing this issue, as the freeze mode fused batch norm has been supported now: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/nn_grad.py#L674

**Comment by DeepLearningJohnDoe** [comment]
Thank you!

**Comment by ppwwyyxx** [comment]
@zhangyaobit Thanks for providing the implementation for backward. However, I noticed that the performance of FusedBatchNormGrad when is_training=False is very pool. 
I didn't benchmark the op alone, but switching is_training on and off can affect my total training speed by 2.5x. My estimate is that the op itself would have at least 5x difference in speed.
I guess the reason is that the implementation now is not based on fused kernels. This makes it as slow as the non-fused kernels.

**Comment by ppwwyyxx** [comment]
@zhangyaobit I used the kernel `BatchNormWithGlobalNormalizationGrad` for the backward implementation when is_training=False. This improves the speed a lot (my overall training time is then closed to the fused case when is_training=True).
To really use this op there are still some issues:
1. The op `BatchNormWithGlobalNormalizationGrad` was marked deprecated. I unmarked it to test but not sure if that's desired. It's a helpful op in this case.
2. The op doesn't support backward. This will break "grad of grad" unit tests. I think "grad of grad" is less important but it could be implemented if really needed.

**Comment by ppwwyyxx** [comment]
I'll send a PR on this.

**Comment by gongbudaizhe** [comment]
Hi @ppwwyyxx ,

I notice that currently only NHWC is supported for FusedBatchNormFreezeGrad and tf.transpose is needed in python if the input is NCHW.  This can take enourmous time if batch norm is adopted in large scale and the data format is NCHW. 

Do you consider adding support for NCHW?

**Comment by ppwwyyxx** [comment]
I don't have time for this. 
In fact a thorough benchmark is needed because the performance of the normal reduction ops has been improved a lot since last year. It's quite likely that my naive cuda kernel isn't very performant in NHWC or NCHW compared to a non-fused multi-op implementation.

**Comment by gongbudaizhe** [comment]
Thanks for your prompt response.

I have done a quick benchmark using my network. For fused BN, it takes ~1.43 second/batch while for non-fused BN, it takes ~1.32 second/batch regardless if NCHW or NHWC is used. It seems for freezed BN, non-fused is faster.

**Comment by ppwwyyxx** [comment]
A quick test shows that fused & non-fused BN have similar backward speed when freezed:
```python
#!/usr/bin/env python
import tensorflow as tf
import time
import os
from tensorflow.python.client import device_lib
local_device_protos = device_lib.list_local_devices()
GPU_MODE = len([x.name for x in local_device_protos if x.device_type == 'GPU'])

N = 64
C = 128
H, W = 128, 128
print("N, C, H, W:", [N, C, H, W])

def benchmark_all(fuse, format):
    shape4d = [N, C, H, W] if format == 'NCHW' else [N, H, W, C]

    tf.reset_default_graph()
    input = tf.get_variable('input', shape=shape4d, dtype=tf.float32)

    scale = tf.get_variable('scale', shape=[C])
    offset = tf.get_variable('offset', shape=[C])
    mean = tf.random_normal(shape=[C])
    var = tf.random_normal(shape=[C]) + 1.

    if fuse:
        output, _, _ = tf.nn.fused_batch_norm(
            input, scale, offset, mean, var, epsilon=1e-5, data_format=format, is_training=False)
    else:
        if format == 'NCHW':
            newshape = [1, C, 1, 1]
            scale = tf.reshape(scale, newshape)
            offset = tf.reshape(offset, newshape)
            mean = tf.reshape(mean, newshape)
            var = tf.reshape(var, newshape)
        output = (input - mean) * (tf.rsqrt(var + 1e-5) * scale) + offset


    forward_op = output.op
    cost = tf.reduce_sum(output)
    backward_op = tf.train.GradientDescentOptimizer(0.1).minimize(cost)


    def benchmark(op, nr_iter=200, nr_warmup=10):
        if not GPU_MODE:
            nr_iter = nr_iter // 10
            nr_warmup = nr_warmup // 5
        for k in range(nr_warmup):
            op.run()
        start = time.perf_counter()
        for k in range(nr_iter):
            op.run()
        end = time.perf_counter()
        itr_per_sec = nr_iter * 1. / (end - start)
        return itr_per_sec

    sess = tf.Session()
    with sess.as_default():
        sess.run(tf.global_variables_initializer())

        spd_forward = benchmark(forward_op)
        print("Fuse={}, Format={}, Forward: {} itr/s".format(fuse, format, spd_forward))
        spd_backward = benchmark(backward_op)
        print("Fuse={}, Format={}, Backward: {} itr/s".format(fuse, format, spd_backward))


if GPU_MODE:
    formats = ['NHWC', 'NCHW']
else:
    formats = ['NHWC']
for format in formats:
    for fuse in [True, False]:
        benchmark_all(fuse, format)
```

Outputs on GTX1080Ti:
```
N, C, H, W: [64, 128, 128, 128]
Fuse=True, Format=NHWC, Forward: 100.62081027587591 itr/s
Fuse=True, Format=NHWC, Backward: 61.392495723794156 itr/s
Fuse=False, Format=NHWC, Forward: 100.51494045143419 itr/s
Fuse=False, Format=NHWC, Backward: 60.875319893753236 itr/s
Fuse=True, Format=NCHW, Forward: 261.46341501158184 itr/s
Fuse=True, Format=NCHW, Backward: 44.5398623696861 itr/s
Fuse=False, Format=NCHW, Forward: 71.64600813406135 itr/s
Fuse=False, Format=NCHW, Backward: 48.042794214933586 itr/s
```
